// dub/resources/links/links.js
"use strict";
// File generated from our OpenAPI spec by Stainless.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Links = void 0;
const resource_1 = require("dub/resource");
const BulkAPI = __importStar(require("dub/resources/links/bulk"));
const InfoAPI = __importStar(require("dub/resources/links/info"));
class Links extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.info = new InfoAPI.Info(this._client);
        this.bulk = new BulkAPI.Bulk(this._client);
    }
    /**
     * Create a new link for the authenticated project.
     */
    create(params, options) {
        const { projectSlug = this._client.projectSlug, ...body } = params;
        return this._client.post('/links', { query: { projectSlug }, body, ...options });
    }
    /**
     * Edit a link for the authenticated project.
     */
    update(linkId, params, options) {
        const { projectSlug = this._client.projectSlug, ...body } = params;
        return this._client.put(`/links/${linkId}`, { query: { projectSlug }, body, ...options });
    }
    /**
     * Retrieve a list of links for the authenticated project. The list will be
     * paginated and the provided query parameters allow filtering the returned links.
     */
    list(params, options) {
        const { projectSlug = this._client.projectSlug, ...query } = params;
        return this._client.get('/links', { query: { projectSlug, ...query }, ...options });
    }
    /**
     * Delete a link for the authenticated project.
     */
    deleteLink(linkId, params, options) {
        const { projectSlug = this._client.projectSlug } = params;
        return this._client.delete(`/links/${linkId}`, { query: { projectSlug }, ...options });
    }
}
exports.Links = Links;
(function (Links) {
    Links.Info = InfoAPI.Info;
    Links.Bulk = BulkAPI.Bulk;
})(Links = exports.Links || (exports.Links = {}));
