// dub/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.fileFromPath = exports.toFile = exports.UnprocessableEntityError = exports.PermissionDeniedError = exports.InternalServerError = exports.AuthenticationError = exports.BadRequestError = exports.RateLimitError = exports.ConflictError = exports.NotFoundError = exports.APIUserAbortError = exports.APIConnectionTimeoutError = exports.APIConnectionError = exports.APIError = exports.DubError = exports.Dub = void 0;
const Core = __importStar(require("./core.js"));
const Errors = __importStar(require("./error.js"));
const Uploads = __importStar(require("./uploads.js"));
const qs = __importStar(require("qs"));
const API = __importStar(require("dub/resources/index"));
/** API Client for interfacing with the Dub API. */
class Dub extends Core.APIClient {
    /**
     * API Client for interfacing with the Dub API.
     *
     * @param {string | undefined} [opts.token=process.env['DUB_API_KEY'] ?? undefined]
     * @param {string | null | undefined} [opts.projectSlug]
     * @param {string} [opts.baseURL=process.env['DUB_BASE_URL'] ?? https://api.dub.co] - Override the default base URL for the API.
     * @param {number} [opts.timeout=1 minute] - The maximum amount of time (in milliseconds) the client will wait for a response before timing out.
     * @param {number} [opts.httpAgent] - An HTTP agent used to manage HTTP(s) connections.
     * @param {Core.Fetch} [opts.fetch] - Specify a custom `fetch` function implementation.
     * @param {number} [opts.maxRetries=2] - The maximum number of times the client will retry a request.
     * @param {Core.Headers} opts.defaultHeaders - Default headers to include with every request to the API.
     * @param {Core.DefaultQuery} opts.defaultQuery - Default query parameters to include with every request to the API.
     */
    constructor({ baseURL = Core.readEnv('DUB_BASE_URL'), token = Core.readEnv('DUB_API_KEY'), projectSlug = null, ...opts } = {}) {
        if (token === undefined) {
            throw new Errors.DubError("The DUB_API_KEY environment variable is missing or empty; either provide it, or instantiate the Dub client with an token option, like new Dub({ token: 'My Token' }).");
        }
        const options = {
            token,
            projectSlug,
            ...opts,
            baseURL: baseURL || `https://api.dub.co`,
        };
        super({
            baseURL: options.baseURL,
            timeout: options.timeout ?? 60000 /* 1 minute */,
            httpAgent: options.httpAgent,
            maxRetries: options.maxRetries,
            fetch: options.fetch,
        });
        this.links = new API.Links(this);
        this.qr = new API.Qr(this);
        this.projects = new API.Projects(this);
        this._options = options;
        this.token = token;
        this.projectSlug = projectSlug;
    }
    defaultQuery() {
        return this._options.defaultQuery;
    }
    defaultHeaders(opts) {
        return {
            ...super.defaultHeaders(opts),
            ...this._options.defaultHeaders,
        };
    }
    authHeaders(opts) {
        return { Authorization: `Bearer ${this.token}` };
    }
    stringifyQuery(query) {
        return qs.stringify(query, { arrayFormat: 'comma' });
    }
}
exports.Dub = Dub;
_a = Dub;
Dub.Dub = _a;
Dub.DubError = Errors.DubError;
Dub.APIError = Errors.APIError;
Dub.APIConnectionError = Errors.APIConnectionError;
Dub.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError;
Dub.APIUserAbortError = Errors.APIUserAbortError;
Dub.NotFoundError = Errors.NotFoundError;
Dub.ConflictError = Errors.ConflictError;
Dub.RateLimitError = Errors.RateLimitError;
Dub.BadRequestError = Errors.BadRequestError;
Dub.AuthenticationError = Errors.AuthenticationError;
Dub.InternalServerError = Errors.InternalServerError;
Dub.PermissionDeniedError = Errors.PermissionDeniedError;
Dub.UnprocessableEntityError = Errors.UnprocessableEntityError;
exports.DubError = Errors.DubError, exports.APIError = Errors.APIError, exports.APIConnectionError = Errors.APIConnectionError, exports.APIConnectionTimeoutError = Errors.APIConnectionTimeoutError, exports.APIUserAbortError = Errors.APIUserAbortError, exports.NotFoundError = Errors.NotFoundError, exports.ConflictError = Errors.ConflictError, exports.RateLimitError = Errors.RateLimitError, exports.BadRequestError = Errors.BadRequestError, exports.AuthenticationError = Errors.AuthenticationError, exports.InternalServerError = Errors.InternalServerError, exports.PermissionDeniedError = Errors.PermissionDeniedError, exports.UnprocessableEntityError = Errors.UnprocessableEntityError;
exports.toFile = Uploads.toFile;
exports.fileFromPath = Uploads.fileFromPath;
(function (Dub) {
    // Helper functions
    Dub.toFile = Uploads.toFile;
    Dub.fileFromPath = Uploads.fileFromPath;
    Dub.Links = API.Links;
    Dub.Qr = API.Qr;
    Dub.Projects = API.Projects;
})(Dub = exports.Dub || (exports.Dub = {}));
exports = module.exports = Dub;
exports.default = Dub;
